File: src/main/org/nlogo/awt/Focus.java
package org.nlogo.awt;

import java.awt.Component;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;

public final strictfp class Focus {

  // this class is not instantiable
  private Focus() { throw new IllegalStateException(); }

  ///

  public static void addNoisyFocusListener(final Component comp) {
    comp.addFocusListener
        (new FocusListener() {
          public void focusGained(FocusEvent fe) {
            System.out.println(comp + " gained focus at " + System.nanoTime());
            System.out.println("oppositeComponent = " + fe.getOppositeComponent());
          }

          public void focusLost(FocusEvent fe) {
            System.out.println(comp + " lost focus at " + System.nanoTime());
            System.out.println("oppositeComponent = " + fe.getOppositeComponent());
          }
        });
  }

}

File: src/main/org/nlogo/awt/Focus.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.awt

import java.awt.Component
import java.awt.event.{ FocusEvent, FocusListener }

object Focus {
  def addNoisyFocusListener(comp: Component) {
    comp.addFocusListener(
      new FocusListener {
        override def focusGained(fe: FocusEvent) {
          println(comp + " gained focus at " + System.nanoTime)
          println("oppositeComponent = " + fe.getOppositeComponent)
        }
        override def focusLost(fe: FocusEvent) {
          println(comp + " lost focus at " + System.nanoTime)
          println("oppositeComponent = " + fe.getOppositeComponent)
        }
      })
  }
}
